src/cmd/ksh93/sh/args.c

```diff
@@ -563,8 +565,16 @@ void sh_applyopts(Shell_t *shp, Shopt_t newflags) {
     if ((!sh_isstate(shp, SH_INIT) && is_privileged) ||
         (sh_isstate(shp, SH_INIT) && is_privileged_off && shp->gd->userid != shp->gd->euserid)) {
         if (!is_option(&newflags, SH_PRIVILEGED)) {
-            setuid(shp->gd->userid);
-            setgid(shp->gd->groupid);
+            while (((status = setuid(shp->gd->userid)) < 0) && (errno == EAGAIN));
```

Comment:
@krader1961 
> Also, In my experience a tight loop like this is dangerous. In general there should be a usleep(100); or similar to rate limit the attempt.

Yes, I abstract this sort of thing (into a new subroutine) and rate-limit the re-try loop
(especially w/ EAGAIN) with an intervening sleep of some sort (actually one of my own).

Just a small note here. Although |usleep(3c)| is multithread-safe on the latest Linux,
it is *not* multithread-safe on many other UNIXi platforms.  The reason: It was never intended to be multithread-safe when it was first introduced.  And many platforms never went back and made it multithread-safe either.  Also, the main UNIXi standards bodies (forget exactly which now, X/Open, ?) have classified |usleep(3c)| as deprecated and not recommended for either new or continued use. They (the standards people) recommend the use of |nanosleep(3rt)| (part of the POSIX "real-time" feature set) instead of |usleep(3c)|.  Another option is to just write one's own (simple enough) using |poll(2)| for millisecond precision. In any event, the idea is to use a form of sleep that is multithread-safe.

Reply:
Good points, @DavidMorano, about `usleep()`. This obviously returns us to the earlier discussion about whether `ksh` should explicitly support multiple threads. I know there are people who argue that additional threads of control might be created without the knowledge of the program and thus every program has to assume it is multi-threaded. I'm not convinced by those arguments. If `ksh` and the APIs it relies on do not explicitly create threads, and its plugin API makes it clear that creating threads is not allowed, it seems okay to me to assume that there will be only one thread.

Having said that I agree that `ksh`, given it supports plugins that it has no control over and which might create threads given the plugin API does not forbid creating threads, has to assume there is more than one thread of execution. And therefore require any API it uses to be thread safe. Which makes me sad since it adds considerable complexity that would otherwise be unnecessary.

Reply:
> Which makes me sad since it adds considerable complexity that would otherwise be unnecessary.

:-)
I have been doing nothing but multithread-safe (and multithreaded) for about 36 years now, but I feel your pain! :-)  There was talk some years ago about moving KSH itself towards a multithread direction (explicitly), but little seemed to come out of that talk. I actually had some misgivings myself about that effort, knowing how difficult getting some things correct within a multithread environment can be. My own feeling is that I would like to see some other constructs within KSH be modernized and refactored before anything w/ internal multithreading moves forward.  But in the end analysis, making programs multithread-safe is not really the hardest thing (by far) out there. But one has to start somewhere.  One can start by making all new code multithread-safe, and then going though the rest of the code to make everything else multithread-safe. It has to start somewhere if one wants to achieve it. The particular difficulty with KSH is that it uses a good bit of the AST library, which itself is riddled with many multithread-UNSAFE subroutines. But the task is not insurmountable. Compiling *out* VMALLOC (itself neither multithread-safe nor fork-safe, both a real requirement for something so fundamental as a "malloc" subsystem) is a very important first step!

For myself, I have large programs (which can function either stand-alone or as ... KSH built-ins!) with unknown numbers of threads (both explicit and dynamically created background threads) performing a myriad of functions and services. The reality is (as I have stated probably several times previously) that all modern code should be written to assume unknown numbers of threads, both explicit threads and library-initiated background threads. I also have deep code stacks of thread-live modules, which even possibly call each other or even themselves in recursive loops. I have KSH built-ins that can call subroutines which eventually call back into other KSH built-ins, some of which might even be a previous built-in that ended up calling itself recursively through intervening layers of subroutines (being used as abstraction interfaces).  Further, I have a lot of dynamically loaded code modules (starting w/ |dlopen(3dl)| and ending w/ |dlclose(3dl)|) that also initiate background threads.  There is some real danger (much more than normally encountered) when using background threads from within dynamically loaded and unloaded modules, especially when all code w/ background threads also has to be fork-safe (some of which I have previously discussed in these spaces). But this is more for the brave of heart out there. Since it should *not* be assumed that KSH built-ins will not be unloaded after being loaded (unloaded like w/ a |dlclose(3dl)|). novice KSH built-in thread users should probably not attempt to create background threads (threads that continue to run past the exit of the built-in and giving control back to KSH proper) from within KSH built-ins. But we are getting ahead of ourselves. Let's start by compiling *out* AST VMALLOC (as I did years ago, but possibly with unknown hidden bugs as @krader1961 discovered); the removal of which being a prerequisite for really any threading direction people might want to take.